components/zcash_address/src/kind/unified.rs

```diff
@@ -193,4 +365,73 @@ mod tests {
             prop_assert_eq!(decoded, Ok(ua));
         }
     }
+
+    #[test]
+    fn duplicate_typecode() {
+        // Construct and serialize an invalid UA.
+        let ua = Address(vec![Receiver::Sapling([1; 43]), Receiver::Sapling([2; 43])]);
+        let encoded = ua.to_bytes();
+        assert_eq!(
+            Address::try_from(&encoded[..]),
+            Err(ParseError::DuplicateTypecode(Typecode::Sapling))
+        );
+    }
+
+    #[test]
+    fn p2pkh_and_p2sh() {
+        // Construct and serialize an invalid UA.
+        let ua = Address(vec![Receiver::P2pkh([0; 20]), Receiver::P2sh([0; 20])]);
+        let encoded = ua.to_bytes();
+        assert_eq!(
+            Address::try_from(&encoded[..]),
+            Err(ParseError::BothP2phkAndP2sh)
+        );
+    }
+
+    #[test]
+    fn only_transparent() {
+        // Encoding of `Address(vec![Receiver::P2pkh([0; 20])])`.
+        let encoded = vec![
+            0x3b, 0x3d, 0xe6, 0xb3, 0xed, 0xaa, 0x0a, 0x36, 0x12, 0xbc, 0x8d, 0x2b, 0x1a, 0xaa,
+            0x27, 0x7e, 0x45, 0xc0, 0xc2, 0x0e, 0xf9, 0x6f, 0x24, 0x9b, 0x79, 0x0a, 0x68, 0x76,
+            0xa8, 0x4c, 0x3f, 0xf0, 0x1f, 0x39, 0x97, 0xbd, 0x15, 0x0d,
+        ];
+
+        // We can't actually exercise this error, because at present the only transparent
+        // receivers we can use are P2PKH and P2SH (which cannot be used together), and
+        // with only one of them we don't have sufficient data for F4Jumble (so we hit a
+        // different error).
+        assert_eq!(
+            Address::try_from(&encoded[..]),
+            Err(ParseError::InvalidEncoding)
+        );
+    }
+
+    #[test]
+    fn receivers_are_sorted() {
+        // Construct a UA with receivers in an unsorted order.
+        let ua = Address(vec![
+            Receiver::P2pkh([0; 20]),
+            Receiver::Orchard([0; 43]),
+            Receiver::Unknown {
+                typecode: 0xff,
+                data: vec![],
+            },
+            Receiver::Sapling([0; 43]),
+        ]);
+
+        // `Address::receivers` sorts the receivers in priority order.
+        assert_eq!(
+            ua.receivers(),
+            vec![
+                Receiver::Orchard([0; 43]),
+                Receiver::Sapling([0; 43]),
+                Receiver::P2pkh([0; 20]),
+                Receiver::Unknown {
+                    typecode: 0xff,
+                    data: vec![],
+                },
+            ]
+        )
+    }
```

Comment:
Add a test for this case:
> Senders MUST reject Unified Addresses/Viewing Keys in which any constituent address does not meet the validation requirements of its Receiver Encoding, as specified in the Zcash Protocol Specification.

An example is a UA containing a Sapling address with a non-canonical encoding (according to ZIP 216) of pk<sub>d</sub>.

Reply:
Also:
> There MUST NOT be additional bytes at the end of the raw encoding that cannot be interpreted as specified above.

Examples would be:
* a UA that ends in a Receiver Encoding with a length field that exceeds the remaining number of bytes;
* a UA where the last Receiver Encoding is truncated after only the Typecode byte.

Reply:
> 
> 
> Add a test for this case:
> 
> > Senders MUST reject Unified Addresses/Viewing Keys in which any constituent address does not meet the validation requirements of its Receiver Encoding, as specified in the Zcash Protocol Specification.
> 
> An example is a UA containing a Sapling address with a non-canonical encoding (according to ZIP 216) of pkd.

`zcash_address` does not currently cover this case, as it only represents receivers using their raw encodings. It would be up to the downstream user of the crate to map these into their own types, and eventually perform those checks. In the case of `zcashd` the earliest we can check this is in the FFI right after this, but even if this were omitted we know it would eventually be rejected by the transaction builder.

Alternatively we could make `zcash_address` depend on all of the underlying crates (i.e. `zcash_primitives`, `orchard`) to parse them, and require the downstream user to have those dependencies in addition to their own.